analysis/monkwindwalker/src/modules/tier/PrimordialPotential.tsx

```diff
@@ -0,0 +1,148 @@
+const PP_MOD = 0.22;
+
+/** Monk Tier 28 "Garb of the Grand Upwelling" 4-Set Bonus for Windwalker */
+class PrimordialPotential extends Analyzer {
+  static dependencies = {
+    abilityTracker: AbilityTracker,
+  };
+  protected abilityTracker!: AbilityTracker;
+
+  /** IDs of abilities casts that was cast during Primordial Power */
+  poweredCasts: { [spellID: number]: number } = {};
+  totalDamage = 0;
+
+  constructor(options: Options) {
+    super(options);
+    this.active = this.selectedCombatant.has4Piece(MONK_TIER_ID);
+    if (!this.active) {
+      return;
+    }
+
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(ABILITIES_AFFECTED_BY_PRIMORDIAL_POWER),
+      this.onCast,
+    );
+    this.addEventListener(
+      Events.damage.by(SELECTED_PLAYER).spell(DAMAGE_AFFECTED_BY_PRIMORDIAL_POWER),
+      this.onDamage,
+    );
+  }
+
+  onCast(event: CastEvent) {
+    if (!this.selectedCombatant.hasBuff(SPELLS.PRIMORDIAL_POWER_BUFF.id)) {
+      return;
+    }
+
+    if (!this.poweredCasts[event.ability.guid]) {
+      this.poweredCasts[event.ability.guid] = 0;
+    }
+
+    this.poweredCasts[event.ability.guid] += 1;
+
+    event.meta = event.meta || {};
+    event.meta.isEnhancedCast = true;
+    const reason = (
+      <>
+        This cast was empowered by <SpellLink id={SPELLS.PRIMORDIAL_POWER_BUFF.id} />
+      </>
+    );
+    event.meta.enhancedCastReason = event.meta.enhancedCastReason ? (
+      <>
+        {event.meta.enhancedCastReason}
+        <br />
+        {reason}
+      </>
+    ) : (
+      reason
+    );
+  }
+
+  onDamage(event: DamageEvent) {
+    if (!this.selectedCombatant.hasBuff(SPELLS.PRIMORDIAL_POWER_BUFF.id)) {
+      return;
+    }
+
+    this.totalDamage += calculateEffectiveDamage(event, PP_MOD);
+  }
+
+  renderCastRatioChart() {
+    const ratioChartItems = ABILITIES_AFFECTED_BY_PRIMORDIAL_POWER.map(
+      (spell: { id: number; name: string }) => ({
+        label: spell.name,
+        spellId: spell.id,
+        color: abilityColors[spell.id] || 'hotpink',
+        value: this.poweredCasts[spell.id] || 0,
+      }),
+    )
+      .filter((item) => item.value > 0)
+      .sort((a, b) => b.value - a.value);
+
+    return (
+      <div className="pad">
+        <div>Breakdown of empowered casts by</div>
+        <DonutChart items={ratioChartItems} />
+      </div>
+    );
+  }
+
+  statistic() {
+    return (
+      <Statistic
+        position={STATISTIC_ORDER.OPTIONAL(2)}
+        size="flexible"
+        category={STATISTIC_CATEGORY.ITEMS}
+        tooltip={
+          <>
+            The {formatPercentage(PP_MOD, 0)}% increase from the 4-set bonus,{' '}
+            <SpellLink id={SPELLS.PRIMORDIAL_POTENTIAL.id} />, was worth{' '}
+            {formatNumber(this.totalDamage)} raw damage.
+          </>
+        }
+        dropdown={this.totalDamage > 0 && this.renderCastRatioChart()}
```

Comment:
Do we really need the `this.totalDamage > 0` aspect of it?

Reply:
I didn't want to add the dropdown if there weren't any spells to show at all, and I don't know how well the donutgraph deals with no items. It was a lot easier to check `this.totalDamage  > 0` than `Object.keys(this.poweredCasts).length > 0`.

Reply:
To me
The vast majority if not all of the logs that get ran through wowA will have at least 13 abilities cast during the course of the encounter meaning we will almost certainty have enough data to show the chart

Reply:
But you can set timespan filters so it's plausible that you manage to filter a period where you never have the primordial power buff so no damage is affected. 

It's a corner case of course, but easy to avoid is my thoguht 😄 Could also just have a `hasEmpoweredCasts = false` and then in the `onCast()`, change it to true and then don't have to check `this.totalDamage` or `this.empoweredCasts`.